methods/SPIA/SPIA.py

```diff
@@ -0,0 +1,292 @@
+
+    """
+    help = __doc__
+    name = "SPIA"
+
+    def __init__(self, organism: str = 'hsa', threshold: float = 0.05, nB: int = 2000, beta=None, markdown: str = ''):
+        """
+
+        Args:
+            organism: organism name (ex. 'Homo sapiens', 'human', 'hsa')
+            threshold: float: threshold for identification of differentially expressed genes
+            nB: number of iterations of random sample choosing at SPIA algorithm
+            beta: list of gene relations values, if None the values are default
+            markdown: generate additional markdown output file with given name
+        """
+        for x in SPECIES:
+            if organism in x:
+                self.organism = x[1]
+                break
+        else:
+            raise Exception("Unknown organism")
+        if threshold < 0 or threshold > 1:
+            raise ValueError('Indices need to be in (0,1) range')
+        self.threshold = threshold
+        self.nB = nB
+        self.beta = beta
+        self.markdown = markdown
+        if markdown:
+            if os.path.exists(markdown if '.md' in markdown else markdown.split('.')[0] + '.md'):
+                print("Warning: '" + markdown + "' file already exists and will be overwritten!")
+
+    @staticmethod
+    def load_data_dict(dictionary, all):
+        '''
+
+        Part of this code is imported from https://github.com/iseekwonderful/PyPathway on MIT license.
+        Load data from certain json.
+
+        Args:
+            dictionary: the json, this method will load data form json
+            all: a python list of total genes. e.g. ['A', 'B', 'C', 'D']
+
+        Returns: the loaded data
+
+        '''
+        data = dictionary
+        Bb = set(all)
+        datpT = {}
+        id2name = data['id2name']
+        for pid, v in data.items():
+            if pid == 'id2name':
+                continue
+            datpT[pid] = {}
+            for key, d in v.items():
+                if key == 'row_names':
+                    for idname in range(len(d)):
+                        Aaa = set(d[idname].split(','))
+                        Ddd = list(Aaa.intersection(Bb))
+                        if Ddd:
+                            if len(Ddd) > 1:
+                                print("Error with gene names in pathway occured")
+                            else:
+                                d[idname] = Ddd[0]
+                    datpT[pid][key] = d
+                else:
+                    m = np.zeros(
+                        (len(v['row_names']), len(v['row_names'])))
+                    for x in d:
+                        m[x[0]][
+                            x[1]] = 1
+                    datpT[pid][
+                        key] = m
+        return datpT, id2name
+
+    @staticmethod
+    def calculate_spia(de, all, dictionary, nB=2000, beta=None, combine='fisher'):
+        """
+        This code is imported from https://github.com/iseekwonderful/PyPathway on MIT license.
```

Comment:
Same as above. NB instead of "imported", I would use "is based upon", as there is no import in there.